import into failed when inject global sort dir network delay 50ms with error “read xxx bytes from external storage, exceed max limit 1073741824”

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、tidb_enable_dist_task='on' and 
2、enable global sort and tidb_cloud_storage_uri is set to minio dir
3、run import into
4、inject minio network delay 50ms

### 2. What did you expect to see? (Required)
import into can success

### 3. What did you see instead (Required)
import failed
						
the status of import job is not finished or running (now: 2024-01-16 02:48:14, jobId: 150003, step: importing, status: failed)
operatorLogs:
[2024-01-16 02:28:44] ###### start import into
[2024-01-16 02:28:44] ###### wait for import job to finish
[2024-01-16 02:48:14] ###### wait for import job to finish failed
select id, step, status from mysql.tidb_import_jobs where start_time >= '2024-01-16 02:28:44'
jobId: 150003, step: importing, status: failed

tidb logs：
[2024/01/16 02:48:06.486 +08:00] [ERROR] [reader.go:49] ["read all data failed"] [takeTime=2m27.255433123s] [error="failed to read file 150003/90036/data/b0dba35c-6041-47e1-af29-cc98ec29d765/40: read 316276121654947278 bytes from external storage, exceed max limit 1073741824"]

[tidb-0.tar.gz](https://github.com/pingcap/tidb/files/13945076/tidb-0.tar.gz)
[tidb-1.tar.gz](https://github.com/pingcap/tidb/files/13945078/tidb-1.tar.gz)


### 4. What is your TiDB version? (Required)
./tidb-server -V
 Release Version: v7.6.0
Edition: Community
Git Commit Hash: 42960c3352d165810c79852b85ada448a68efa4c
Git Branch: heads/refs/tags/v7.6.0
UTC Build Time: 2024-01-15 11:33:46
GoVersion: go1.21.5
Race Enabled: false
Check Table Before Drop: false
Store: unistore
2024-01-16T02:28:41.835+0800

another error for this scenario
`[2024/04/21 07:08:31.840 +08:00] [ERROR] [task_executor.go:248] ["failed to handle task"] [task-id=120008] [task-type=ImportInto] [error="failed to read file 120008/120046/data/2027856e-b25a-4c79-a7b8-1879de5793e8/0: illegal n (-1256825232400295505) when reading from external storage"]`

--- a/br/pkg/storage/ks3.go
+++ b/br/pkg/storage/ks3.go
@@ -450,6 +450,7 @@ func (rs *KS3Storage) Open(ctx context.Context, path string, o *ReaderOption) (E
 		storage:      rs,
 		name:         path,
 		reader:       reader,
+		pos:          r.Start,
 		rangeInfo:    r,
 		prefetchSize: prefetchSize,
 	}, nil

--- a/br/pkg/storage/s3.go
+++ b/br/pkg/storage/s3.go
@@ -828,6 +828,7 @@ func (rs *S3Storage) Open(ctx context.Context, path string, o *ReaderOption) (Ex
 		storage:      rs,
 		name:         path,
 		reader:       reader,
+		pos:          r.Start,
 		ctx:          ctx,
 		rangeInfo:    r,
 		prefetchSize: prefetchSize,

--- a/br/pkg/storage/s3_test.go
+++ b/br/pkg/storage/s3_test.go
@@ -14,6 +14,7 @@ import (
 	"os"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"testing"
 
 	"github.com/aws/aws-sdk-go/aws"
@@ -854,6 +855,52 @@ func (s *s3Suite) expectedCalls(ctx context.Context, t *testing.T, data []byte, 
 		}
 		lastCall = thisCall
 	}
+}
+
+type mockFailReader struct {
+	r         io.Reader
+	failCount *atomic.Int32
+}
+
+func (f *mockFailReader) Read(p []byte) (n int, err error) {
+	if f.failCount.Load() > 0 {
+		f.failCount.Add(-1)
+		return 0, errors.New("mock read error")
+	}
+	return f.r.Read(p)
+}
+
+func TestS3RangeReaderRetryRead(t *testing.T) {
+	s := createS3Suite(t)
+	ctx := aws.BackgroundContext()
+	content := []byte("0123456789")
+	var failCount atomic.Int32
+	s.s3.EXPECT().GetObjectWithContext(gomock.Any(), gomock.Any()).
+		DoAndReturn(func(_ context.Context, input *s3.GetObjectInput, opt ...request.Option) (*s3.GetObjectOutput, error) {
+			var start int
+			_, err := fmt.Sscanf(*input.Range, "bytes=%d-", &start)
+			require.NoError(t, err)
+			requestedBytes := content[start:]
+			return &s3.GetObjectOutput{
+				Body:         io.NopCloser(&mockFailReader{r: bytes.NewReader(requestedBytes), failCount: &failCount}),
+				ContentRange: aws.String(fmt.Sprintf("bytes %d-%d/%d", start, len(content)-1, len(content))),
+			}, nil
+		}).Times(2)
+	reader, err := s.storage.Open(ctx, "random", &ReaderOption{StartOffset: aws.Int64(3)})
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, reader.Close())
+	}()
+	slice := make([]byte, 2)
+	n, err := reader.Read(slice)
+	require.NoError(t, err)
+	require.Equal(t, 2, n)
+	require.Equal(t, []byte("34"), slice)
+	failCount.Store(1)
+	n, err = reader.Read(slice)
+	require.NoError(t, err)
+	require.Equal(t, 2, n)
+	require.Equal(t, []byte("56"), slice)
 }
 
 // TestS3ReaderWithRetryEOF check the Read with retry and end with io.EOF.

